### ISSUE TITLE
NullValueMappingStrategy with Lists

### ISSUE THREAD

**Issue by kevcodez** [mapstruct/mapstruct#1273]
Given the following example

```
public class Dto {

    List<Long> longs = new ArrayList<>();

    public List<Long> getLongs() {
        return longs;
    }

    public void setLongs(List<Long> longs) {
        this.longs = longs;
    }
}
```

```
public class Entity {

    List<Long> longs = new ArrayList<>();

    public List<Long> getLongs() {
        return longs;
    }

    public void setLongs(List<Long> longs) {
        this.longs = longs;
    }
}
```

```
@Mapper(nullValueMappingStrategy = NullValueMappingStrategy.RETURN_DEFAULT)
public interface EntityMapper {

    Dto asTarget(Entity entity);

}
```

The generated mapper looks like this:

```public class EntityMapperImpl implements EntityMapper {
    public EntityMapperImpl() {
    }

    public Dto asTarget(Entity entity) {
        Dto dto = new Dto();
        if (entity != null) {
            List<Long> list = entity.getLongs();
            if (list != null) {
                dto.setLongs(new ArrayList(list));
            } else {
                dto.setLongs((List)null);
            }
        }

        return dto;
    }
}
```

Shouldn't the list set to an empty list rather than null? Or do I have a misunderstanding of the null value mapping strategy?

**Comment by filiphr** [comment]
I think that you might be right...

Although in the JavaDoc it is written:

>If `null` is passed to a mapping method, a default value will be returned. The value depends on the kind of the annotated method:

Where **passed** is the key word. Maybe it should be considered for the other locations as well.

**Comment by kevcodez** [comment]
> However, by specifying nullValueMappingStrategy = NullValueMappingStrategy.RETURN_DEFAULT on @BeanMapping, @IterableMapping, @MapMapping, or globally on @Mapper or @MappingConfig, the mapping result can be altered to return empty default values. This means for:
> Iterables / Arrays: an empty iterable will be returned. 
> The strategy works in a hierarchical fashion. Setting nullValueMappingStrategy on mapping method level will override @Mapper#nullValueMappingStrategy, and @Mapper#nullValueMappingStrategy will override @MappingConfig#nullValueMappingStrategy.

Since the strategy can also be used on an IterableMapping or set globally on a Mapper, I expected the list to be empty rather than null. I am not sure if this is working as intended for now, or if this is a bug.

**Comment by filiphr** [comment]
As I said, I think that you are right and we need to set an empty collection there. I'll mark this as a bug. 

I think that your first option is something similar as in my first suggestion of #879

**Comment by kevcodez** [comment]
@filiphr Should this be part of the current NullValueMappingStrategy? Or should this be another separate setting?

**Comment by filiphr** [comment]
@kevcodez what part do you want to work on?

I think that the correct solution for this issue would be the second option if `NullValueMappingStrategy#RETURN_DEFAULT`is used then we need to set it to an empty collection. There is no need for a new strategy for this.

If you are talking about the #879 issue that is something completely different and should be discussed there.

**Comment by kevcodez** [comment]
@filiphr I just want to fix this issue. Ok, understood, not sure if I am able to provide a PR in the next days after work, hopefully be able to look into it.

**Comment by filiphr** [comment]
@kevcodez sure. If you want I can point you to the location that generates that part of the code.

**Comment by kevcodez** [comment]
@filiphr I'd appreciate that

**Comment by filiphr** [comment]
The wrapper that generate code like that are all the wrapper that extend `WrapperForCollectionsAndMaps`. The ftl files for those wrappers are: `ClassName.ftl`. If you need more help let me know

**Comment by kevcodez** [comment]
[_1273.zip](https://github.com/mapstruct/mapstruct/files/1273254/_1273.zip)

I only had a few minutes so far, only wrote a test case to reproduce the bug. All classes can be copy and pasted into the **mapstruct\processor\src\test\java\org\mapstruct\ap\test\bugs**.

@filiphr There is no ClassName.ftl, also I did not quickly find the right spot to edit. Might have to take a deeper look

**Comment by filiphr** [comment]
@kevcodez I meant the name of the classes that extend the `WrapperForCollectionsAndMaps`. Sorry for not making myself clear enough

**Comment by kevcodez** [comment]
@filiphr The SetterWrapperForCollectionsAndMapsWithNullCheck.ftl contains

```java
<@lib.handleExceptions>
  <@callTargetWriteAccessor/>
  <#if !ext.defaultValueAssignment??>else {<#-- the opposite (defaultValueAssignment) case is handeld inside lib.handleLocalVarNullCheck -->
    ${ext.targetBeanName}.${ext.targetWriteAccessorName}<@lib.handleWrite>null</@lib.handleWrite>;
  }
  </#if>
</@lib.handleExceptions>
```

The `<@lib.handleWrite>null</@lib.handleWrite>` seems to be the part to change, in case the NullValueMappingStrategy is set to **RETURN_DEFAULT**. I would most likely implement the change in the related java class. However, I would like to discuss on the implementation detail first. Depending on the type used, the assignment needs to differ, e.g.

`List` -> `new ArrayList`
`Map` -> `new HashMap`
`Set` -> `new HashSet`
`Collection` -> ? No-Op? new List?
`Iterable` -> Same as Collection

Which types should be supported and should MapStruct just assume the implemention class to be used for the most common case? WDYT?

**Comment by filiphr** [comment]
Regarding the type to use, we already have support for that.

Have a look at [`IterableCreation.ftl`](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl#L28). Those types that you are mentioning already have an `implementationType`, so you can use that to create it.

If you think that we have types for which an implementation type is missing let us know, that should be a bug and should be part of another issue. The implementation types are defined [here](http://mapstruct.org/documentation/stable/reference/html/#implementation-types-for-collection-mappings) in our documentation. That part is handled when the `Type` itself is created, so you shouldn't worry too much about it.

**Comment by filiphr** [comment]
Closed with https://github.com/mapstruct/mapstruct/commit/499dbd4561305c747f869013c717137dfe6a5fbc

**Comment by filiphr** [comment]
Just an information about this issue. 

The approach that we took to implement this turned out to not be quite "correct", therefore as part of PR https://github.com/mapstruct/mapstruct/pull/1618 and issue https://github.com/mapstruct/mapstruct/issues/1306 we decided that using `NullValueMappingStrategy#RETURN_DEFAULT` will not affect how properties values are set. Therefore we have added `NullValuePropertyMappingStrategy#SET_TO_DEFAULT` that can be used on `@MapperConfig`, `@Mapper`, `@BeanMapping` and `@Mapping` to control how the property should be mapped. This now only affects update methods. Which means that for normal mappings one should either set the defaults inside the object creation or do them via `@ObjectFactory`. The last bit was done in order to stay consistent with how mappings are handled by other properties. 

For the example in this issue no setter would be invoked.
